asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py

```diff
@@ -0,0 +1,926 @@
+    """
+    Creates a string unique identifier for each installer company, since multiple company names might
+    refer to the same installer.
+    E.g. "Company A Ltd", "Company A Limited", "Company A (old account), "Company A t/a Company B"
+    all refer to the same company which will be identified by the unique ID "company a".
+
+    Identifiers are created by transforming the original company names:
+    - Cleaning the company name (strings to lowe case, removing stopwords and punctuation);
+    - Merge all different "trading as" expressions together;
+    - Creating a dictionary to map companies trading as other companies to be identified as the same;
+    - Creating a dictionary to map companies with slightly different names with address in the same location;
+
+    Args:
+        installer_data: historical installers data
+    """
+    # Cleaning company name
+    installer_data["processed_company_name"] = installer_data["company_name"].apply(
+        clean_company_name
+    )
+
+    # Dealing with different versions of "trading as"
+    installer_data["processed_company_name"] = installer_data[
+        "processed_company_name"
+    ].apply(deal_with_versions_of_trading_as)
+
+    # Unique companies containing "trading as" in the name
+    trading_as_companies = installer_data[
+        installer_data["processed_company_name"].str.contains("trading as")
+    ][["processed_company_name"]]
+    trading_as_companies.drop_duplicates("processed_company_name", inplace=True)
+
+    # Creating dicitionary to map "trading as" companies to the same company name
+    trading_as_dictionary = dictionary_mapping_trading_as_company_names(
+        trading_as_companies
+    )
+
+    # Applying the mapping created above to create the company unique ID
+    installer_data["company_unique_id"] = installer_data[
+        "processed_company_name"
+    ].apply(
+        lambda x: trading_as_dictionary[x] if x in trading_as_dictionary.keys() else x
+    )
+
+    # Dropping processed_company_name as no longer needed
+    installer_data.drop("processed_company_name", axis=1, inplace=True)
+
+    # Companies have the same location but not currently identified as being the same
+    same_postcode_no_match = installer_data.groupby("postcode", as_index=False)[
+        ["company_unique_id"]
+    ].nunique()
+    same_postcode_no_match = same_postcode_no_match[
+        same_postcode_no_match["company_unique_id"] > 1
+    ]["postcode"].unique()
+
+    # New unique ID matches based on companies with same location
+    new_matches = map_installers_same_location_different_id(
+        same_postcode_no_match, installer_data
+    )
+    installer_data["company_unique_id"] = installer_data["company_unique_id"].apply(
+        lambda x: new_matches[x] if x in new_matches.keys() else x
+    )
+
+
+# copied from process_mcs_utils.py and changed slightly
+def geocode_postcode(data: pd.DataFrame, geodata: pd.DataFrame) -> pd.DataFrame:
```

Comment:
I'm adding the coordinates as this was also previously done by India in the work she did on the installers snapshot data, and it might be useful for a few analyses. Just trying to make this as broad as possible, to account for other analyses besides HPMT. :)